feat: enable user to enter custom prompt

AmanNegi opened the pull request:
## Description

- Usually normal users would be relying on the LLM for prompt generation but power users who wish to enter their own prompt an now mention `p` or `prompt` as the first argument and pass their `prompt` as the second argument.
- This provides the power user a finer control over the prompt and in turn the output.

## Preview

![Screenshot from 2024-06-25 14-18-33](https://github.com/RocketChat/Apps.AI.GIF/assets/37607224/419039d4-ee25-4091-945f-554e0fe488a1)
> The power user can preview the GIF generation directly in the preview section.

Fixes #10 

@Nabhag8848 please let me know if this seems fine and if any changes are required.

Nabhag8848 commented:
@AmanNegi  lemme know once you resolve this conflicts

AmanNegi committed:
chore: fix merge conflicts

Nabhag8848 commented on src/commands/GenGifCommand.ts:
```diff
@@ -107,30 +125,62 @@ export class GenGifCommand implements ISlashCommand {
-                context
-            );
+            if (args == "p" || args == "prompt") {
+                args = context.getArguments()[1].trim();
```
@AmanNegi can you implement commandUtitlity class now for this. Remember i mention it earlier ?

AmanNegi replied:
Yep it's the right time to add the utitliy class, also later we could compile the history into this.  Got that done!

Nabhag8848 commented on src/lib/GifRequestDispatcher.ts:
```diff
@@ -170,4 +171,78 @@ export class GifRequestDispatcher {
+        let output: string | undefined;
+
+        while (
+            status !== PredictionStatus.SUCCEEDED &&
+            status !== PredictionStatus.FAILED &&
+            status !== PredictionStatus.CANCELLED
```
Instead of have this you can use includes, Following: 

```suggestion
 ```
    const invalidStatuses = [
    PredictionStatus.SUCCEEDED,
    PredictionStatus.FAILED,
    PredictionStatus.CANCELLED]; 
     while(!invalidStatuses.includes(status)) {
       // Your code here
     }

Nabhag8848 reviewed (changes requested):
@AmanNegi  Added few comments, lemme know once you are done.

AmanNegi committed:
chore: implement command utility classes

- includes minor refactoring changes, and simplification of
  GenGifCommand class.

Nabhag8848 commented on definition/command/ICommandUtility.ts:
```diff
@@ -0,0 +1,59 @@
+    room: IRoom;
+    read: IRead;
+    modify: IModify;
+    http: IHttp;
+    persis: IPersistence;
+    triggerId?: string;
+    threadId?: string;
+    requestDebouncer: RequestDebouncer;
+
+    resolveCommand(): Promise<ISlashCommandPreview>;
+    resolveExecutePreviewItem(item: ISlashCommandPreviewItem): Promise<void>;
+}
+
+export interface IPreviewerUtilityParams {
+    app: AiGifApp;
+    params: Array<string>;
+    sender: IUser;
+    room: IRoom;
+    read: IRead;
+    modify: IModify;
+    http: IHttp;
+    persis: IPersistence;
+    triggerId?: string;
+    threadId?: string;
+    requestDebouncer: RequestDebouncer;
+}
+
+export interface IPreviewItemUtilityParams {
+    app: AiGifApp;
+    params: Array<string>;
+    item: ISlashCommandPreviewItem;
+    sender: IUser;
+    room: IRoom;
+    read: IRead;
+    modify: IModify;
+    http: IHttp;
+    persis: IPersistence;
+    triggerId?: string;
+    threadId?: string;
+}
```
@AmanNegi typescript has utility types which lets us reuse types/interface properties.
Refer: https://www.typescriptlang.org/docs/handbook/utility-types.html play around with it, use those and come up with something reusable.

AmanNegi replied:
Interesting, I find `Omit` particularly useful here. :sparkle: :rocket:

Nabhag8848 reviewed (changes requested):
@AmanNegi

AmanNegi committed:
nit: add type reusability

Nabhag8848 commented on definition/command/ICommandUtility.ts:
```diff
@@ -0,0 +1,50 @@
+    ISlashCommandPreviewItem,
+} from "@rocket.chat/apps-engine/definition/slashcommands";
+
+export interface ICommandUtility {
+    app: AiGifApp;
+    params: Array<string>;
+    sender: IUser;
+    room: IRoom;
+    read: IRead;
+    modify: IModify;
+    http: IHttp;
+    persis: IPersistence;
+    triggerId?: string;
+    threadId?: string;
+    requestDebouncer: RequestDebouncer;
+
+    resolveCommand(): Promise<ISlashCommandPreview>;
+    resolveExecutePreviewItem(item: ISlashCommandPreviewItem): Promise<void>;
+}
+
+export interface IPreviewerUtilityParams {
+    app: AiGifApp;
+    params: Array<string>;
+    sender: IUser;
+    room: IRoom;
+    read: IRead;
+    modify: IModify;
+    http: IHttp;
+    persis: IPersistence;
+    triggerId?: string;
+    threadId?: string;
+    requestDebouncer: RequestDebouncer;
+}
```
@AmanNegi we can apply here as well in some way, try that out.

AmanNegi replied:
Okay I'll simplify it, i don't believe we have the need for all those params in the `ICommandUtility` class. 

> Edit: After discussion I understood how removing it undermines the use of interfaces. Further, I will work on implementing it in a proper way.

Nabhag8848 reviewed (changes requested):
@AmanNegi

AmanNegi committed:
nit: simplify CommandUtilityClass

AmanNegi committed:
nit: add var types to ICommandUtility

AmanNegi committed:
nit: update types

Nabhag8848 reviewed (approved):
@AmanNegi  LGTM
